src/EditorFeatures/Core/Implementation/KeywordHighlighting/HighlighterViewTaggerProvider.cs

```diff
@@ -3,74 +3,100 @@
 using Microsoft.VisualStudio.Text;
 using Microsoft.VisualStudio.Text.Editor;
 using Microsoft.VisualStudio.Text.Tagging;
 using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Editor.Implementation.Highlighting
 {
     [Export(typeof(IViewTaggerProvider))]
-    [TagType(typeof(HighlightTag))]
+    [TagType(typeof(KeywordHighlightTag))]
     [ContentType(ContentTypeNames.CSharpContentType)]
     [ContentType(ContentTypeNames.VisualBasicContentType)]
     [TextViewRole(PredefinedTextViewRoles.Interactive)]
-    internal class HighlighterViewTaggerProvider :
-        AbstractAsynchronousViewTaggerProvider<HighlightTag>
+    internal class HighlighterViewTaggerProvider : AsynchronousTaggerProvider<KeywordHighlightTag>
     {
-        private readonly IHighlightingService _highlighterService;
+        private readonly IHighlightingService _highlightingService;
+
+        // Whenever an edit happens, clear all highlights.  When moving the caret, preserve 
+        // highlights if the caret stays within an existing tag.
+        public override TaggerCaretChangeBehavior CaretChangeBehavior => TaggerCaretChangeBehavior.RemoveAllTagsOnCaretMoveOutsideOfTag;
+        public override TaggerTextChangeBehavior TextChangeBehavior => TaggerTextChangeBehavior.RemoveAllTags;
+        public override IEnumerable<Option<bool>> Options => SpecializedCollections.SingletonEnumerable(InternalFeatureOnOffOptions.KeywordHighlight);
 
         [ImportingConstructor]
         public HighlighterViewTaggerProvider(
+            IHighlightingService highlightingService,
             IForegroundNotificationService notificationService,
-            IHighlightingService highlighterService,
             [ImportMany] IEnumerable<Lazy<IAsynchronousOperationListener, FeatureMetadata>> asyncListeners)
             : base(new AggregateAsynchronousOperationListener(asyncListeners, FeatureAttribute.KeywordHighlighting), notificationService)
         {
-            _highlighterService = highlighterService;
+            _highlightingService = highlightingService;
         }
 
-        protected override bool RemoveTagsThatIntersectEdits => true;
-
-        protected override SpanTrackingMode SpanTrackingMode => SpanTrackingMode.EdgeExclusive;
-
-        protected override IEnumerable<Option<bool>> TagSourceOptions
-        {
-            get
-            {
-                yield return InternalFeatureOnOffOptions.KeywordHighlight;
-            }
-        }
-
-        protected override ITaggerEventSource CreateEventSource(ITextView textView, ITextBuffer subjectBuffer)
+        public override ITaggerEventSource CreateEventSource(ITextView textView, ITextBuffer subjectBuffer)
         {
             return TaggerEventSources.Compose(
-                TaggerEventSources.OnTextChanged(subjectBuffer, TaggerDelay.OnIdle, reportChangedSpans: true),
+                TaggerEventSources.OnTextChanged(subjectBuffer, TaggerDelay.OnIdle),
```

Comment:
If we have a tree changed one, this would be another place it technically shoud be used. Or worst case, also update to listen to parse options changed and we'll clean it up when we get rid of that.